Implement missing features and fix found bugs for BigQuery Client Library

vam-google opened the pull request:
0) Access query results row cells by name. Now `row.get(1)` and `get("firstColumnName")` are supported (currently supported only for results returned by `BigQuery.query()` and not supported by results returned `BigQuery.listTableData()` because there is no schema and it would require additional request to the server to get schema (but still can be added in the future easily with cost of an extra network call)). Same syntax is supported for record-type fields (i.e. nested fields can be accessed by name too).
1) Refactor `BigQuery.query()` to use `jobs.insert/jobs.getQueryResults` combination instead of `jobs.query`. `QueryRequest` class was removed, now `query()` method directly accepts `QueryJobConfiguration` instead. The remaining properties of removed `QueryRequest` class now are passed in a form of `QueryOption` vararg argument. `QueryOption` combines query and waiting options (waiting options are necessary because query() is now waiting for query to complete).
2) Make `BigQuery.query()` call synchronous (waits for query completion before return result), with jittered exponential backoff for polling (based on `gax.retrying` package)
3) Add client side job id generation (`JobId.of()`)
4) Replace `WaitForOption` with `RetryOption` (based on `RetrySettings` and is consistent with the rest of the codebase). This affected compute and spanner clients too.
5) Rewrite `Job.wait()` to use jittered exponentiall backoff polling (based on`gax.retrying` package). Polling is performed differently depending on the type of job: for query jobs on each poll `getQueryResults()` is called, for all other job types `getJob()` is used instead.
6) Use standard SQL as default for all queries
7) Fix wrong query results iteration samples code used all over the place in documentation.
8) Various smaller changes/refactorings

vam-google commented:
@vkedia please review spanner's `Operation.waitFor()`, as it was rewrittent as part of this PR.

vam-google commented on google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/QueryResponse.java:
```diff
@@ -16,6 +16,7 @@
 
 package com.google.cloud.bigquery;
 
+import com.google.cloud.bigquery.BigQuery.QueryOption;
```
QueryOption is used in documentation

pongad commented on google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQuery.java:
```diff
@@ -440,6 +443,72 @@ public static QueryResultsOption maxWaitTime(long maxWaitTime) {
+      return option instanceof RetryOption ? (RetryOption) option : null;
+    }
+
+    public static QueryResultsOption[] filterQueryResutlsOptions(QueryOption... options) {
```
"Results"?

Why arrays? I'd think lists are more idiomatic?

vam-google replied:
Yes, we should use collections as return types instead of arrays if possible. In this particular case arrays are used because our manual clients use Options in a form of vararg arguments (someMethod(QueryResultsOption... options), so the output of this method is most likely to be passed to a method, which accepts a vararg argument of QueryResultsOption, and we can pass an array as a vararg argument (and it would work as expected). If it was a list we would have to convert it to an array first. So I decided to return array directly to avoid unnecessary conversions.

pongad replied:
ah, sgtm

tswast commented on google-cloud-examples/src/main/java/com/google/cloud/examples/bigquery/snippets/BigQuerySnippets.java:
```diff
@@ -571,22 +572,18 @@ public boolean cancelJobFromId(String jobName) {
-      response = bigquery.getQueryResults(response.getJobId());
-    }
+    QueryJobConfiguration queryConfig =
+        QueryJobConfiguration.newBuilder(query).setUseLegacySql(true).build();
```
Ideally the runQuery example would show a standard SQL sample and we'd have a separate sample for explicitly setting to legacy SQL.

vam-google replied:
I agree, but I believe this should be made as a separate PR. The goal of this is to modify the bigquery client itself (changing snippets only if they are broken).

tswast commented on google-cloud-examples/src/main/java/com/google/cloud/examples/bigquery/snippets/BigQuerySnippets.java:
```diff
@@ -596,25 +593,20 @@ public QueryResponse runQuery(String query) throws InterruptedException {
     // [START runQueryWithParameters]
-    QueryRequest request = QueryRequest.newBuilder(query)
+    QueryJobConfiguration queryConfig = QueryJobConfiguration.newBuilder(query)
         .setUseLegacySql(false) // standard SQL is required to use query parameters
```
No longer required to explicitly set legacy to false now, correct?

vam-google replied:
Yes. I removed the setter but left and edited the corresponding comment (to emphasize that parameters work only with Std SQL).

tswast commented on google-cloud-examples/src/main/java/com/google/cloud/examples/bigquery/snippets/InsertDataAndQueryTable.java:
```diff
@@ -79,20 +83,16 @@ public static void main(String... args) throws InterruptedException {
-            .setPageSize(1000L)
-            .build();
+    QueryJobConfiguration queryConfig =
+        QueryJobConfiguration.newBuilder("SELECT * FROM my_dataset_id.my_table_id").build();
```
We should not be encouraging people to [use `select *` with BigQuery](https://stackoverflow.com/a/23991604).

vam-google replied:
I did not know that. What should be used instead? Explicit list of column names?
I did not create that query, only modified the surrounding code, that is how that line is showed as "added". Leaving it as is.

tswast replied:
`SELECT my_column FROM my_dataset_id.my_table_id` would be appropriate.

vkedia commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/Operation.java:
```diff
@@ -120,43 +135,57 @@
+                if (!(prevThrowable instanceof SpannerException)) {
+                  return ((BaseServiceException) prevThrowable).isRetryable();
+                }
+                return ((SpannerException) prevThrowable).getErrorCode() != ErrorCode.NOT_FOUND;
```
I don't think this is correct. We should only retry if SpannerException.isRetryable is true otherwise we should just bubble up the exception.

vam-google replied:
Seems reasonable, but the old code was doing this:

        if (e.getErrorCode() == ErrorCode.NOT_FOUND) {
          return null;
        }
        if (!e.isRetryable()) {
          throw e;
        }

I.e. it was essentially terminating in case if the error code was not found, and this is what is done here.

vkedia replied:
In the old code, if the exception is SpannerException and is retryable, it would retry. Here if the exception is SpannerException and is not NOT_FOUND it would retry since it returns true in that case.

vam-google replied:
Yes, you are right. I reversed the up the order of condition checks, because I thought it would make code shorter. But it just made it wrong =).

garrettjonesgoogle commented on google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/Field.java:
```diff
@@ -245,16 +98,31 @@ public Builder setName(String name) {
-    public Builder setType(Type type) {
-      this.type = checkNotNull(type);
-      return this;
+    public Builder setType(LegacySQLTypeName type, Field... subFields) {
```
It seems weird to require a "Legacy" SQL type name for non-legacy use cases.

vam-google replied:
100% agree. Though it is an older problem, just made more visible with this PR. Before, there was a wrapper type, which "encapsulated" legacy type (basically one more wrapper around it, but it did nto have word "Legacy" in it). But the wrapper type was still using the legacy terminology (for example RECORD (legacy) instead of STRUCT (std)).

So I just removed a wrapper which was essentially doing nothing, it was not hiding the legacy nature of the types, since the legacy terminology was used. 

Currently we have LegacySQLTypeName referencing std types, so when you know legacy type, you can easy find a mapped std type. But it doesn't work in the opposite direction.

As we discussed offline switching to StandardSQLTypeName can be done in separate PR

garrettjonesgoogle commented on google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/Fields.java:
```diff
@@ -0,0 +1,83 @@
+    ImmutableMap.Builder<String, Integer> nameIndexBuilder = ImmutableMap.builder();
+    int index = 0;
+    for (Field field : fields) {
+      nameIndexBuilder.put(field.getName(), index++);
```
I much prefer the increment to be on its own line.

vam-google replied:
Well, I like using inline incrementing (feels like "true programming" :), but ok, changed.

garrettjonesgoogle commented on google-cloud-bigquery/src/test/java/com/google/cloud/bigquery/FieldValueTest.java:
```diff
@@ -50,75 +50,75 @@
   @Test
   public void testFromPb() {
-    FieldValue value = FieldValue.fromPb(BOOLEAN_FIELD);
+    FieldValue value = FieldValue.fromPb(BOOLEAN_FIELD, null);
```
If it's common to pass `null` as a second parameter, do we want to create a `fromPb` overload that only requires the first parameter?

vam-google replied:
So I was thinking about same thing it during making this change. I'll add the overloaded version.

vam-google commented on google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryImpl.java:
```diff
@@ -473,19 +468,21 @@ public Rows apply(RowToInsert rowToInsert) {
       String cursor = result.x();
       return new PageImpl<>(new TableDataPageFetcher(tableId, serviceOptions, cursor, optionsMap),
-          cursor, transformTableData(result.y()));
+          cursor, transformTableData(result.y(), null));
```
Theoretically yes, but the current implementation is based on retrieving the actual schema. With record field types and multiple layers of schema nesting the things get complicated (that is why a real schema is needed).

vkedia commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/Operation.java:
```diff
@@ -120,43 +135,57 @@
+                return !prevResponse.getDone();
+              }
+              if (prevThrowable instanceof BaseServiceException) {
+                if (!(prevThrowable instanceof SpannerException)) {
```
why are you checking for this not to be instance of SpannerException?

vam-google replied:
Oh, I think the idea was to "support" other retriable exceptions, which theoretically can be added in the future. 

But you are right, because of that I did it wrong, and all spanner exceptions will be retried to oblivion regardless if they are retriable or no. I've almost introduced a a time bomb bug =). Thanks!

vam-google commented:
Addressed review feedback. PTAL.

@vkedia I fixed the retry logic (sorry for for breaking it). Hopefully now it is correct (matches exactly the original one in functionality). Please note that it theoretically may refuse to retry a retriable BaseServiceException, which is retriable but is not a SpannerException. This is only a theoretical concern, since all BaseServiceExceptions in this context will always be SpanneExceptions too.
Also, please note that with this change all unknown unchecked exceptions will also be wrapped in SpannerException (different from the original code). I think it makes sense here. Please let me know if that should be changed too.

garrettjonesgoogle commented on google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQuery.java:
```diff
@@ -440,6 +443,72 @@ public static QueryResultsOption maxWaitTime(long maxWaitTime) {
+      return option instanceof RetryOption ? (RetryOption) option : null;
+    }
+
+    public static QueryResultsOption[] filterQueryResutlsOptions(QueryOption... options) {
```
spelling (Resutls)

Also why is it necessary for this to be public?

vam-google replied:
Fixed the typo. It doesn't have to be public, but I thought it would not hurt to be public. Changed to package-private.

garrettjonesgoogle commented on google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQuery.java:
```diff
@@ -939,74 +994,80 @@ public static QueryResultsOption maxWaitTime(long maxWaitTime) {
+   * String query = "SELECT distinct(corpus) FROM `bigquery-public-data.samples.shakespeare`";
+   * QueryJobConfiguration queryConfig = QueryJobConfiguration.of(query);
+   *
+   * // To run the legacy syntax queries use the following code instead:
```
Also doesn't the query have slightly different syntax too?

vam-google replied:
Yes, updated the comment with corresponding legacy syntax.

garrettjonesgoogle commented on google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/FieldValues.java:
```diff
@@ -0,0 +1,114 @@
+ * may or may not contain related schema. If schema is not provided, the individual cells of the row
+ * will still be accessible by index but not by name.
+ */
+public class FieldValues extends AbstractList<FieldValue> implements Serializable {
```
It seems like this should be called `FieldValueList`

vam-google replied:
I like FieldValues and Fields more, as they are shorter and still clear. But I changed to FieldList and FieldValueList to make clear that those are valid lists.

garrettjonesgoogle commented on google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/Fields.java:
```diff
@@ -0,0 +1,118 @@
+ * Google BigQuery Table schema fields (columns). Each field has a unique name and index. Fields
+ * with duplicate names are not allowed in BigQuery schema.
+ */
+public final class Fields extends AbstractList<Field> implements Serializable {
```
It seems like this should be called `FieldList`

vam-google replied:
I like FieldValues and Fields more, as they are shorter and still clear. But I changed to FieldList and FieldValueList to make clear that those are valid lists.

garrettjonesgoogle commented:
LGTM
